Make the key ring cache expire shortly after an immediately-activated key is generated

amcasey opened the pull request:
If a key is required and none is available, one will be generated and immediately activated.  Immediately-activated keys are problematic because they may be used before they have propagated to other app instances.  What's more, in a scenario where one instance needs to generate an immediately-activated key, it's likely that other instances will need to do the same.  A given instance has no control over when other instances pick up its immediately-activated keys, but it can do its best to pick up theirs.  Therefore, every time it generates an immediately-activated key, it assumes other instances may have as well and pre-emptively schedules a refresh of its own key ring cache.

This problem is already partially (and complementarily) addressed by `KeyRingProvider.InAutoRefreshWindow`, which allows missing keys to be re-fetched from the backing repository for two minutes after app-startup.  This will cover the case where the short expiration period is too short to catch all updates but doesn't help with cases where an immediately-activated key is required after initial startup.

Sadly, this still does not completely address the issue: instances could still use keys unknown to their peers during the 15 second refresh window.  To fix the issue properly, we'd need to know the order in which the keys had been persisted to the repository.

Part of #52678, which is part of #36157
Should benefit from #54490

amcasey committed:
Make the key ring cache expire shortly after an immediately-activated key is generated

If a key is required and none is available, one will be generated and immediately activated.  Immediately-activated keys are problematic because they may be used before they have propagated to other app instances.  What's more, in a scenario where one instance needs to generate an immediately-activated key, it's likely that other instances will need to do the same.  A given instance has no control over when other instances pick up its immediately-activated keys, but it can do its best to pick up theirs.  Therefore, every time it generates an immediately-activated key, it assumes other instances may have as well and pre-emptively schedules a refresh of its own key ring cache.

This problem is already partially (and complementarily) addressed by `KeyRingProvider.InAutoRefreshWindow`, which allows missing keys to be re-fetched from the backing repository for two minutes after app-startup.  This will cover the case where the short expiration period is too short to catch all updates but doesn't help with cases where an immediately-activated key is required after initial startup.

Sadly, this still does not completely address the issue: instances could still use keys unknown to their peers during the 15 second refresh window.  To fix the issue properly, we'd need to know the order in which the keys had been persisted to the repository.

Part of #52678, which is part of #36157

adityamandaleeka commented on src/DataProtection/DataProtection/src/KeyManagement/KeyManagementOptions.cs:
```diff
@@ -117,6 +117,20 @@ internal static TimeSpan MaxServerClockSkew
+    /// of its own key ring cache.  This property controls how long after an immediately-
+    /// activated key is generated the key ring cache will be refreshed.
+    /// </summary>
+    internal static TimeSpan ShortKeyRingRefreshPeriod { get; set; } = TimeSpan.FromSeconds(15);
```
nit: maybe put this adjacent to KeyRingRefreshPeriod

amcasey commented on src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.Tests/KeyManagement/KeyRingProviderTests.cs:
```diff
@@ -787,6 +788,158 @@ public void GetCurrentKeyRing_WithExpiredExistingKeyRing_UpdateFails_ThrowsButCa
+    }
+
+    [Fact]
+    public async Task MultipleInstanceGenerateImmediatelyActivatedKeys()
```
"Instances"

amcasey commented:
@adityamandaleeka I'm tempted to add an appcontext switch for this.  Thoughts?

amcasey commented on src/DataProtection/DataProtection/src/KeyManagement/KeyRingProvider.cs:
```diff
@@ -164,7 +164,24 @@ private CacheableKeyRing CreateCacheableKeyRingCoreStep2(DateTimeOffset now, Can
+            // (in which case, other instances may have done the same)
+            ? (generatedKey.ActivationDate < now + KeyManagementOptions.KeyPropagationWindow) // No clock skew on a key we generated
+            // Or we selected a key that has yet to propagate (presumably, from another instance)
+            : (defaultKey.CreationDate > now - KeyManagementOptions.KeyPropagationWindow);
```
Can this happen in a loop?  What if this is still the default after the next refresh?

amcasey replied:
Indeed it can. 😢

amcasey commented:
Idea: what if, instead of refreshing the key ring, we entered/extended an auto-refresh window.

We'd have no guarantee of an unknown key triggering a refresh within the window, so probably not ideal.

amcasey committed:
Move ShortKeyRingRefreshPeriod near KeyRingRefreshPeriod

amcasey committed:
Fix typo

adityamandaleeka commented:
> @adityamandaleeka I'm tempted to add an appcontext switch for this. Thoughts?

Sounds good to me. Under a "new data protection" switch that includes other stuff I assume? Feels like a separate switch for this might be overkill.

amcasey committed:
Update test to assert we don't get stuck in a short-refresh loop

amcasey committed:
Avoid back-to-back short refresh periods

We don't want to refresh every 15 seconds until the propagation window elapses.  Storing the flag on the CacheableKeyRing simplifies thread-safe state management, but requires changes to some public test hooks.  This version of the change should be source- and binary-compatible with previous versions of .net.

amcasey committed:
Decouple ICacheableKeyRingProvider2 from ICacheableKeyRingProvider

amcasey committed:
Rename ICacheableKeyRingProvider2 to IInternalCacheableKeyRingProvider

...to parallel IInternalXmlKeyManager.

amcasey committed:
Merge branch 'main' into StartupRace

captainsafia commented on src/DataProtection/DataProtection/src/KeyManagement/KeyManagementOptions.cs:
```diff
@@ -76,6 +76,20 @@ internal static TimeSpan KeyRingRefreshPeriod
+    /// of its own key ring cache.  This property controls how long after an immediately-
+    /// activated key is generated the key ring cache will be refreshed.
+    /// </summary>
+    internal static TimeSpan ShortKeyRingRefreshPeriod { get; set; } = TimeSpan.FromSeconds(15);
```
Any guidance here on what the minimum refresh period should be or why 15 seconds is a sane default?

captainsafia commented on src/DataProtection/DataProtection/src/KeyManagement/KeyRingProvider.cs:
```diff
@@ -167,7 +170,25 @@ private CacheableKeyRing CreateCacheableKeyRingCoreStep2(DateTimeOffset now, Can
+        // of downtime may discover that it has a valid, but soon-to-be-expired key. The replacement will not
+        // be immediately-activated, but may be activated before it has propagated.
+        var useShortRefreshPeriod = allowShortRefreshPeriod &&
+            (generatedKey is not null
```
Nit: Move the second clause of this Boolean to a well-defined method (e.g. `GeneratedKeyWithinPropogationWindow` or something).

captainsafia commented on src/DataProtection/DataProtection/src/KeyManagement/KeyRingProvider.cs:
```diff
@@ -241,7 +262,7 @@ internal IKeyRing GetCurrentKeyRingCore(DateTime utcNow, bool forceRefresh = fal
                 try
                 {
-                    newCacheableKeyRing = CacheableKeyRingProvider.GetCacheableKeyRing(utcNow);
+                    newCacheableKeyRing = CacheableKeyRingProvider.GetCacheableKeyRing(utcNow, allowShortRefreshPeriod: existingCacheableKeyRing?.HasShortRefreshPeriod != true);
```
What does `existingCacheableKeyRing?.HasShortRefreshPeriod != true` functionally mean here?

amcasey replied:
It's a nullable bool

amcasey replied:
AppContext switch here?

amcasey commented:
In the absence of races (e.g. as in the deterministic simulator), this provides essentially no benefit and slightly increases the amount of work done (e.g. network calls).  Since races tend to occur on startup and we already have a startup mitigation (the auto-reload period), this change probably does not justify the complexity it adds.

amcasey commented:
@captainsafia Your feedback is still welcome, but this is now low priority.
